### PR TITLE
feat(ingest/mongodb): support stateful ingestion

### PR DESCRIPTION
**Pull request by TonyOuyangGit**
This PR improves the MongoDB ingestion source by adding stateful ingestion and platform instance support and also refactors the existing code to emit `MetadataChangeProposal` instead of `MetadataChangeEvent` which is the current standard way of emitting events.

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/mongodb.py [hunk 1]
```diff
@@ -90,6 +103,10 @@ class MongoDBConfig(PlatformInstanceConfigMixin, EnvConfigMixin):
     # mongodb only supports 16MB as max size for documents. However, if we try to retrieve a larger document it
     # errors out with "16793600" as the maximum size supported.
     maxDocumentSize: Optional[PositiveInt] = Field(default=16793600, description="")
+    ingest_data_platform_instance_aspect: Optional[bool] = Field(
+        default=False,
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/mongodb.py:107]: we probably want this to be default true, or at least done automatically if platform_instance is set. don't think it makes sense to make it a config option

File: metadata-ingestion/src/datahub/ingestion/source/mongodb.py [hunk 2]
```diff
@@ -59,7 +72,7 @@
 DENY_DATABASE_LIST = set(["admin", "config", "local"])
 
 
-class MongoDBConfig(PlatformInstanceConfigMixin, EnvConfigMixin):
+class MongoDBConfig(DatasetSourceConfigMixin, StatefulIngestionConfigBase):
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/mongodb.py:75]: While it's true that DatasetSourceConfigMixin just combines the other two
```
class DatasetSourceConfigMixin(PlatformInstanceConfigMixin, EnvConfigMixin):
```

we're generally trying to move away from tall config inheritance hierarchies, and would prefer wide hierarchies that are more explicit and understandable

that's a long way of saying that I'd prefer if it was 
```suggestion
class MongoDBConfig(PlatformInstanceConfigMixin, EnvConfigMixin, StatefulIngestionConfigBase):
```

**Comment by hsheth2** [comment]
@TonyOuyangGit one thing - platform_instance support was already added in https://github.com/datahub-project/datahub/commit/852267972c8efc1ceb5a0cbd71594d7ea2529d49 - it looks like this is just doing some refactoring which is great, but I'm going to tweak the PR title accordingly

**Review by hsheth2 — APPROVED** [review]

**Comment by hsheth2** [comment]
@TonyOuyangGit looks like the mongodb tests need to be updated

**Comment by TonyOuyangGit** [comment]
@hsheth2 Thank you! Since I refactored to use `MetadataChangeProposal` to emit datasets, I think I'll need to update the golden file to reflect the changes.

**Pull request merged**